lib/api/objectDelete.js

```diff
@@ -24,48 +30,132 @@ export default function objectDelete(authInfo, request, log, cb) {
+                (err, options) => {
+                    if (err) {
+                        log.debug('err processing versioning info',
+                        { error: err });
+                        return next(err, bucketMD);
+                    }
+                    return next(null, bucketMD, objectMD, options);
                 });
-                return cb(err, corsHeaders);
-            }
-            if (!objMD) {
-                return cb(errors.NoSuchKey, corsHeaders);
+        },
+        function deleteOperation(bucketMD, objectMD, delOptions, next) {
+            const deleteInfo = {
+                removeDeleteMarker: false,
+                newDeleteMarker: false,
+            };
+            if (delOptions && delOptions.deleteData) {
+                if (objectMD.isDeleteMarker) {
+                    // record that we deleted a delete marker to set
+                    // response headers accordingly
+                    deleteInfo.removeDeleteMarker = true;
+                }
+                return services.deleteObject(bucketName, objectMD, objectKey,
+                    delOptions, log, (err, delResult) => next(err, bucketMD,
+                    objectMD, delResult, deleteInfo));
             }
-            const headerValResult = validateHeaders(objMD, request.headers);
-            if (headerValResult.error) {
-                return cb(headerValResult.error, corsHeaders);
+            // putting a new delete marker
+            deleteInfo.newDeleteMarker = true;
+            return createAndStoreObject(bucketName, bucketMD,
+                objectKey, objectMD, authInfo, canonicalID, null, request,
+                deleteInfo.newDeleteMarker, null, log, (err, newDelMarkerRes) =>
+                next(err, bucketMD, objectMD, newDelMarkerRes, deleteInfo));
+        },
+    ], (err, bucketMD, objectMD, result, deleteInfo) => {
+        const resHeaders = collectCorsHeaders(request.headers.origin,
+            request.method, bucketMD);
+        if (err) {
+            log.debug('error processing request', { error: err,
+                method: 'objectDelete' });
+            return cb(err, resHeaders);
+        }
+        if (deleteInfo.newDeleteMarker) {
+            // if we created a new delete marker, return true for
+            // x-amz-delete-marker and the version ID of the new delete marker
+            if (result.versionId) {
+                resHeaders['x-amz-delete-marker'] = true;
+                resHeaders['x-amz-version-id'] = result.versionId === 'null' ?
+                    result.versionId : versionIdUtils.encrypt(result.versionId);
             }
-            if (objMD['content-length']) {
-                log.end().addDefaultFields({
-                    contentLength: objMD['content-length'],
-                });
+            // TODO metric for delete marker
+            pushMetric('putObject', log, { authInfo, bucket: bucketName,
```

Comment:
@ttvi Interesting, I would assume the number of objects count would reflect what you would see from list objects api and not list object versions api. Let me play with S3 metrics to see how they are handling this situation.